interpreter/cling/lib/Interpreter/ExternalInterpreterSource.cpp

```diff
@@ -0,0 +1,302 @@
+//------------------------------------------------------------------------------
+
+#include "ExternalInterpreterSource.h"
+#include "cling/Interpreter/Interpreter.h"
+
+#include "clang/Lex/Preprocessor.h"
+
+using namespace clang;
+
+namespace {
+  class ClingASTImporter : public ASTImporter {
+  private:
+  cling::ExternalInterpreterSource& m_Source;
+
+  public:
+    ClingASTImporter(ASTContext &ToContext, FileManager &ToFileManager,
+                     ASTContext &FromContext, FileManager &FromFileManager,
+                     bool MinimalImport, cling::ExternalInterpreterSource& source):
+      ASTImporter(ToContext, ToFileManager, FromContext, FromFileManager,
+                  MinimalImport), m_Source(source) {}
+
+    Decl *Imported(Decl *From, Decl *To) override {
+      ASTImporter::Imported(From, To);
+
+      if (isa<TagDecl>(From)) {
+        clang::TagDecl* toTagDecl = dyn_cast<TagDecl>(To);
+        toTagDecl->setHasExternalLexicalStorage();
+        toTagDecl->setMustBuildLookupTable();
+      }
+
+      if (isa<NamespaceDecl>(From))
+      {
+          NamespaceDecl *toNamespaceDecl = dyn_cast<NamespaceDecl>(To);
+          toNamespaceDecl->setHasExternalVisibleStorage();
+      }
+     
+      if (isa<ObjCContainerDecl>(From))
+      {
+          ObjCContainerDecl *toContainerDecl = dyn_cast<ObjCContainerDecl>(To);
+          toContainerDecl->setHasExternalLexicalStorage();
+          toContainerDecl->setHasExternalVisibleStorage();
+      }
+
+      // Put the name of the Decl imported with the
+      // DeclarationName coming from the parent, in  my map.
+      if (NamedDecl *fromNamedDecl = llvm::dyn_cast<NamedDecl>(From)) {
+        NamedDecl *toNamedDecl = llvm::dyn_cast<NamedDecl>(To);
+        m_Source.addToImportedDecls(toNamedDecl->getDeclName(),
+                                    fromNamedDecl->getDeclName());
+      }
+      if (DeclContext *fromDeclContext = llvm::dyn_cast<DeclContext>(From)) {
+        DeclContext *toDeclContext = llvm::dyn_cast<DeclContext>(To);
+        m_Source.addToImportedDeclContexts(toDeclContext, fromDeclContext);
+      }
+
+      return To;
+    }
+  };
+}
+
+namespace cling {
+
+  ExternalInterpreterSource::ExternalInterpreterSource(const cling::Interpreter *parent,
+                                              cling::Interpreter *child) :
+    m_ParentInterpreter(parent), m_ChildInterpreter(child) {
+
+    clang::DeclContext *parentTUDeclContext =
+      clang::TranslationUnitDecl::castToDeclContext(
+      m_ParentInterpreter->getCI()->getASTContext().getTranslationUnitDecl());
+
+    clang::DeclContext *childTUDeclContext =
+      clang::TranslationUnitDecl::castToDeclContext(
+      m_ChildInterpreter->getCI()->getASTContext().getTranslationUnitDecl());
+
+    // Also keep in the map of Decl Contexts the Translation Unit Decl Context
+    m_ImportedDeclContexts[childTUDeclContext] = parentTUDeclContext;
+  }
+
+  ClingASTImporter CreateClingASTImporter(ASTContext &toContext,
```

Comment:
Ok if they are fixed regardless of the childCurrentContext then sure I can make them data memebrs :)